### ISSUE TITLE
Make Matlab species thermo properties consistent and useful

### ISSUE THREAD

**Issue by speth** [Cantera/cantera#522]
The documentation for the Matlab function `enthalpies_RT` says it returns the "standard-state" species enthalpies. However, it ultimately calls the C++ method `ThermoPhase::getEnthalpy_RT_ref`, which returns what Cantera calls the "reference state" enthalpies. This represents the confluence of several problems:

* The name of the function does not indicate that it is returning values for the standard state rather than the current state
* No function exists in the Matlab toolbox to return the actual species enthalpies
* The terminology of "reference state" vs "standard state" is inconsistent between the Matlab toolbox and the C++ base, with the Matlab terminology (I think) more closely matching the more common thermodynamic convention.

The same issue applies for `gibbs_RT`, `entropies_R`, and `Cp_R`, except that for these, the documentation doesn't even say that the values returned are for the reference state.

**Comment by decaluwe** [comment]
Hi Ray,

I've been thinking about your third bullet, some more:

The complex part is that there are two relevant "standard states:"  

1. A standard state for reference species properties.  This is the one that most people think about, and what you identify as the "common thermodynamic convention": species properties are readily calculated for a given temperature and reference pressure, assuming that X_k = 1.0 for a given species.  This could also be called the "reference state" in that is it literally the state for which you have known species properties which you can reference.

2. However, there is also a *phase* "standard state."  This distinct from and cannot be equal to the species reference state, as it refers to the mixture properties at a given state (including composition), and there is obviously no state for which X_k = 1 for all species.  This state is frequently required for derivations many thermodynamic equations of state.  It is a state at which *phase* properties (e.g. mixture-averaged properties) can be readily calculated.  For a non-ideal EoS, for example, departure functions can be calculated by integrating from the "standard state" to the actual state.  

So the two are distinct and necessarily so.  The situation is easily confused because of the simplicity of the ideal gas EoS, where no "standard phase state" is actually required.  One can calculate all thermodynamic properties quite readily, directly from the reference species properties, and mixture rules do not require any complex mathematics.  So it's very easy (and quite common) to use the two terms synonymously.

That said, do we need both (or either) "standard state" and "reference species property" calculations featured in the codebase?  You can't really avoid having the reference data loaded and stored as such, but I'm somewhat skeptical that we need the standard state data explicitly calculated or stored.  

1. The whole point of the standard state is that its properties are readily calculated.  So why not just plug these expressions into the departure function calculations?  Explicitly storing both reference and standard information is needlessly confusing.  
2. I don't really see the "standard state" values being anything that one would with to access for subsequent calculations, beyond what is already done in the class itself.
3. Also, explicitly calculating the standard state info separate from the departure functions is perhaps dangerous, if someone uses the current class as a basis for a new derived class (i.e. the two cannot be modified independently).

As I'm typing this, an additional confusion just reared its ugly head:  the `standardConcentrations` used in the kinetic calculations are also wrapped up in this confusion.  More specifically, the species activity concentrations are `gamma_k*X_k/vbar`, where `gamma_k` is the activity coefficient and `vbar` the mixture-averaged molar volume.  

The form of `gamma_k` comes from the EoS: when one writes the chemical potential for a species k, it can be represented as:
mu_k = mu_k_o + RT ln(a_k) = mu_k_o + RT ln(gamma_k*X_k/vbar/C_k_o)

where C_k_o are the `standardConcentrations`, and mu_k_o are the species chemical potentials at the standard concentrations.  There doesn't really seem to be any clear "correct" choice as to whether this is a "standard" state or a "reference" state:

1. On the one hand, the activity coefficients come from the EoS, so it would make sense that the same "standard state" used to derive the EoS is that used to calculate the activity coefficients.
2. On the other hand, this is clearly a "species property" calculation, and the mu_k_o will obviously involve the "reference" species data (regardless of whether they perform additional calculations).
3. The difference is all semantics, in this case: I just checked the `gamma_k` calculations for the `RedlichKwongMFTP` and the (in development) `PengRobinsonMFTP` classes, and the choice of this standard state has no effect on the calculated activity coefficient.  So the choice of C_k_o simply affects the value of mu_k_o.  

But obviously, it would be nice to get some clarity, here.  I just don't yet know what that should be.  I'm leaning toward just having the "reference" state being the only one referenced in the code, which seems to be your preference, as well.  I'm just still a little wary that this may have unintended consequences or cause other unanticipated confusions.

**Comment by speth** [comment]
A status update for the "experimental" Matlab toolbox: The `ThermoPhase.enthalpies_RT` property ultimately calls `ThermoPhase::getEnthalpy_RT_ref`, as before. However, there are no properties to replace the old `gibbs_RT`, `entropies_R`, or `Cp_R` methods. I think this needs to be addressed in order to complete https://github.com/Cantera/enhancements/issues/177, whether or not we ever settle on a consistent definition of what "standard state" means.

**Comment by ischoegl** [comment]
For the sake of consistency, this issue would be resolved by providing the same interface for MATLAB as for Python, correct? In the case of MATLAB's `ThermoPhase.enthalpies_RT`, Python uses `ThermoPhase.standard_enthalpies_RT` (while calling _different_ C++ functions on the backend, i.e. `ThermoPhase::getEnthalpy_RT_ref` and `ThermoPhase::getEnthalpy_RT`).
 
The distinctions of the corresponding docstrings aren't necessarily clear to me with
https://github.com/Cantera/cantera/blob/9d0e54a3ac1b0532fc1317d34b32a51fce5b8da8/include/cantera/thermo/ThermoPhase.h#L944-L951
vs.
https://github.com/Cantera/cantera/blob/9d0e54a3ac1b0532fc1317d34b32a51fce5b8da8/include/cantera/thermo/ThermoPhase.h#L866-L872
with experimental MATLAB
https://github.com/Cantera/cantera/blob/9d0e54a3ac1b0532fc1317d34b32a51fce5b8da8/interfaces/matlab_experimental/Base/ThermoPhase.m#L122-L133
and Python
https://github.com/Cantera/cantera/blob/9d0e54a3ac1b0532fc1317d34b32a51fce5b8da8/interfaces/cython/cantera/thermo.pyx#L1664-L1668
The legacy MATLAB toolbox stated
```
%     Vector of standard-state species enthalpies
%     divided by RT, where R is the universal gas
%     constant and T is the temperature. For gaseous species, these
%     values are ideal gas enthalpies.
```
which indicates some copy-pasta issue in the new toolbox docstring.

Overall, is this just about one using a reference pressure while the other isn't? I.e. MATLAB references data that are typically provided in textbook appendices (e.g. Appendix A, Glassman 3rd ed)?

What is the relationship to what is calculated here? Is this what MATLAB uses?
https://github.com/Cantera/cantera/blob/9d0e54a3ac1b0532fc1317d34b32a51fce5b8da8/include/cantera/thermo/SpeciesThermoInterpType.h#L191-L205

The other properties mentioned on top would be `ThermoPhase.standard_gibbs_RT`, `ThermoPhase.standard_entropies_R` and `ThermoPhase.standard_cp_R` in the Python API.

Given that we have a new _experimental_ MATLAB toolbox, there's a window of opportunity to make things consistent.

_PS: one alternative would be to switch the Python API to the MATLAB convention by dropping the `standard_` prefix?_

_PPS: this is code from (now removed) `PDSS_IdealGas.cpp`, which may shed some light on this? I.e. the `ss` presumably refers to "standard-state"?_
```C++
void PDSS_IdealGas::setTemperature(double temp)
{
    m_temp = temp;
    m_spthermo->updatePropertiesTemp(temp, &m_cp0_R, &m_h0_RT, &m_s0_R);
    m_g0_RT = m_h0_RT - m_s0_R;
    m_V0 = GasConstant * m_temp / m_p0;
    m_hss_RT = m_h0_RT;
    m_cpss_R = m_cp0_R;
    m_sss_R = m_s0_R - log(m_pres/m_p0);
    m_gss_RT = m_hss_RT - m_sss_R;
    m_Vss = GasConstant * m_temp / m_pres;
}
```

**Comment by ischoegl** [comment]
Alright. Some further digging cleared things up (especially in conjunction with @decaluwe's comments). With definitions
https://github.com/Cantera/cantera/blob/9d0e54a3ac1b0532fc1317d34b32a51fce5b8da8/include/cantera/thermo/PDSS.h#L457-L467
and, similarly 
https://github.com/Cantera/cantera/blob/9d0e54a3ac1b0532fc1317d34b32a51fce5b8da8/include/cantera/thermo/VPStandardStateTP.h#L259-L261
and
https://github.com/Cantera/cantera/blob/9d0e54a3ac1b0532fc1317d34b32a51fce5b8da8/include/cantera/thermo/VPStandardStateTP.h#L278-L280
as well as
https://github.com/Cantera/cantera/blob/9d0e54a3ac1b0532fc1317d34b32a51fce5b8da8/src/thermo/VPStandardStateTP.cpp#L46-L50
and
https://github.com/Cantera/cantera/blob/9d0e54a3ac1b0532fc1317d34b32a51fce5b8da8/src/thermo/VPStandardStateTP.cpp#L99-L103
everything follows quite nicely.

Re-reading @decaluwe's comments, I tend to agree that it may not make sense to provide an API for standard-state quantities, which I believe (?) I read from

> The situation is easily confused because of the simplicity of the ideal gas EoS, where no "standard phase state" is actually required. One can calculate all thermodynamic properties quite readily, directly from the reference species properties, and mixture rules do not require any complex mathematics. So it's very easy (and quite common) to use the two terms synonymously.

> [...] do we need both (or either) "standard state" and "reference species property" calculations featured in the codebase? You can't really avoid having the reference data loaded and stored as such, but I'm somewhat skeptical that we need the standard state data explicitly calculated or stored.

and 

> I don't really see the "standard state" values being anything that one would with to access for subsequent calculations, beyond what is already done in the class itself.

The crux is that `standard_*` quantities are used in one Python sample (`coverage_dependent_surf.py`) as well as several tests. Better documentation with consistent nomenclature, in combination with consistent Python and MATLAB API's may be the best we can do?

_PS: here's further context (sorry if I am documenting this here as I am reading up to this)_
https://github.com/Cantera/cantera/blob/9d0e54a3ac1b0532fc1317d34b32a51fce5b8da8/include/cantera/thermo/ThermoPhase.h#L178-L200

_In conclusion, I'd be in favor to not expose the first two levels (reference state and standard state) at advanced APIs, i.e. deprecate `standard_*` quantities in Python, and introduce `partial_molar_*` quantities in MATLAB._

**Comment by speth** [comment]
I'm 👍 on leaving the "reference state" properties out of the high-level APIs (and perhaps, they should even be `protected` members of the C++ `ThermoPhase` class) as they are an implementation detail.

The "standard state" properties are not as important as the "partial molar" ones, but I think there are some cases where they can be useful to look at. Of course, this runs into the perpetual issue of Cantera's definition of "standard state" being different from the ones generally used in thermodynamics textbooks.

The other source of variability in here that doesn't really help is the mix of dimensional versus nondimensional quantities, for example providing nondimensional reference state but dimensional partial molar properties. Likewise, here, I think it would be best to regard the nondimensional properties as implementation details. It's probably not worth immediately removing such accessors, but at least we can avoid increasing the proliferation of these functions.
